### ISSUE TITLE
Implement an autoloader

### ISSUE THREAD

**Issue by ericnicolaas** [Charitable/Charitable#404]
The primary benefit of this would be to reduce memory consumption, force greater consistency of file naming/organization and simplify charitable.php somewhat.

**Comment by helgatheviking** [comment]
I like this idea, but I think you may have to rename a _lot_ of classes to get this to work.

**Comment by ericnicolaas** [comment]
Yep, pretty sure you're right :)

**Comment by helgatheviking** [comment]
If you rename everything, I don't know if you can avoid breaking a lot of stuff. I'm trying to learn about namespaces and letting this idea marinate.

**Comment by helgatheviking** [comment]
So today I learned a little about autoloading and namespacing. I tried to write an autoloader and by itself that part wasn't so bad, but for it to work, you have to add a namespace *every* class _and_ then use fully-qualified class names everywhere. So every `new Charitable_Some_Admin_Class()` needs to be modified to `new \Charitable\Admin\Charitable_Some_Admin_Class()` and every `Charitable_Emails::get_instance()` needs to be switched to `\Charitable\Emails\Charitable_Emails::get_instance()`. 

I spent about a day doing that and *still* kept running into errors, so this isn't an easy process. And that's without even renaming the classes to remove the `Charitable_` prefix or rename the files to match. 

The best I could come up with with an autoloader that has a giant array that maps the class name to a file. I will send a PR just to show you what I did. That is not quite a true autoloader, but the current class name system doesn't tell you anything about where the file may be found. And as I found renaming the classes so that they *do* tell you where the file is is both really hard *and* bound to break a ton of sites.

**Comment by helgatheviking** [comment]
I did have the thought that if you renamed all the classes then you could perhaps have backcompatibility by including all the old classes as deprecated, but extending the new class.

for example:

```
class Charitable_Some_Admin_Class extends \Charitable\Admin\Charitable_Some_Admin_Class {
 // add deprecation notice here
}
```

but like I said, renaming all the functions is a slow, tedious, process.

**Comment by ericnicolaas** [comment]
Closed with 615b9d8c3b79488e8baca7e88d0f19f62aae93bb

Thanks for this @helgatheviking!
